### ISSUE TITLE
Array::to_bytes has no contiguity check: silently returns the backing buffer's leading bytes for a strided view

### ISSUE THREAD

**Issue by Pushkinist** [Pushkinist/rMLX#473]
## Summary

`Array::to_bytes()` reads `mlx_array_nbytes` bytes straight off the raw data pointer with no contiguity check. On a strided view — a slice, a transpose — that returns the *backing buffer's* leading elements under the *view's* shape. It is silent: no error, no length mismatch, no crash. The byte count is the logical one, so every downstream length check passes.

**Nothing is broken today.** All 96 production callers are currently safe. But roughly half are safe *by accident* — laundered by an `astype` copy, or by a tensor rank that happens to make the slice contiguous — not by any established contiguity property. Two independent single-condition margins are all that separate this from silent KV corruption, and one of them is a property this codebase has repeatedly lost. Filing as a latent hazard with the census as evidence.

## The hazard, quoted

`crates/rmlx-mlx/src/lib.rs:988-1008` (derived-from-source):

```rust
pub fn to_bytes(&self) -> Result<Vec<u8>> {
    install_error_handler();
    self.eval()?;
    let nbytes = unsafe { sys::mlx_array_nbytes(self.inner) };
    if nbytes == 0 { return Ok(Vec::new()); }
    let ptr = unsafe { sys::mlx_array_data_uint8(self.inner) };
    ...
    let bytes = unsafe { std::slice::from_raw_parts(ptr, nbytes) }.to_vec();
    Ok(bytes)
}
```

`nbytes` is `size() * itemsize()` — the *logical* element count. The data pointer absorbs the view's byte offset but not its strides. `eval()` materialises the compute graph; it does not relayout. MLX's slice and transpose primitives share the parent buffer and adjust strides rather than copying, so an evaluated view still has the parent's layout.

`array_to_f32_vec` (`crates/rmlx-kv-quant/src/kvcache/helpers.rs:196-211`) forwards the array unrelaid on the `Dtype::F32` branch and inherits the hazard:

```rust
let a_f32 = if a.dtype() == Dtype::F32 {
    a.try_clone()?              // <- no relayout; hazard inherited
} else {
    a.astype(Dtype::F32, device)?   // <- real copy; launders the view
};
a_f32.eval()?;
let bytes = a_f32.to_bytes()?;
```

The `astype` branch is what makes every non-f32 caller safe. That is the accident.

### Measured

Reproduced on CPU against the linked MLX (mlx 0.31.2 / mlx-c 0.6.0_2), in a throwaway crate outside the repo — no engine code touched:

| case | result |
|---|---|
| `[1,2,4,2]` f32, `.slice(axis-2 prefix → [1,2,2,2])` | `to_bytes` → `[0,1,2,3,4,5,6,7]`; correct is `[0,1,2,3,8,9,10,11]` — **silently wrong** |
| same view, `.contiguous()` first | `[0,1,2,3,8,9,10,11]` — correct |
| `[1,2,4,2]` f32, `.transpose(&[0,1,3,2])` | returns the untransposed buffer verbatim — **silently wrong** |
| `.reshape()` of a non-contiguous view | correct — MLX relayouts on reshape when the input is not row-contiguous |
| arch V shape `[1,S,H,D].transpose(0,2,1,3)`, **S=1** | contiguous — correct |
| same, **S=3** | **non-contiguous — silently wrong** |

The last two rows are the load-bearing ones; see "How close this is" below.

The reshape row is a useful lemma: any caller reading a `reshape` output is safe regardless of what fed the reshape.

## Census

97 `.to_bytes()` call sites outside `*_tests.rs` / `tests.rs` / `tests/` (193 more in tests). Two of those 97 are not `Array::to_bytes` calls — a doc-comment line in `crates/rmlx-kv-ssd/src/block_io.rs:2458` and a `CStr::to_bytes()` at `crates/rmlx-mlx/src/nax.rs:324`. One more, `crates/rmlx-kv-quant/src/test_utils.rs:172`, is test support. **96 production call sites.**

Plus 66 helper call sites that reach `to_bytes` indirectly — 44 `array_to_f32_vec` + 22 `arrays_to_f32` (each of the latter is two reads) — **all** in `crates/rmlx-kv-quant/src/kvcache/update.rs`.

| class | n | basis | verdict |
|---|---|---|---|
| Reduction / elementwise / matmul / kernel output | ~45 | `argmax`, `max_axis`, `tanh`, `sigmoid`, `matmul`, MSL kernel outputs all allocate a fresh dense result | **SAFE** |
| `reshape` output (`logits_flat` family) | ~12 | measured: MLX relayouts on reshape of a non-contiguous input | **SAFE** |
| `astype` to a differing dtype | 3 | real copy; dense result | **SAFE** |
| safetensors hydrate side (`block_io.rs` 2468–3203) | 19 | arrays built from mapped file buffers; contiguous by construction | **SAFE** |
| rank-1 leading-prefix slice | 6 | `slice_prefix` (`quant_k_gpu_ring.rs:710`), paged `gather` (`paged/alloc.rs:141`), `*_trimmed` in `block_io.rs` (989/1045/1121/1168/1223/1289). A rank-1 stride-1 slice is always contiguous — the offset lands in the data pointer | **SAFE (by rank)** |
| SSD spill write side (`OwnedTensor::from_array`, `block_io.rs:169`) | 1 | guarded — see below | **SAFE (deliberately)** |
| Misc: empty-array eval forcing, `CStr`, test support | 4 | `arch/loader.rs:255` hits the `nbytes == 0` early return | **SAFE / n-a** |
| `array_to_f32_vec` F32 branch, reached from `update.rs` | 66 helper sites | **safe only because the KV stream is bf16/f16, never f32** — see below | **CONDITIONALLY SAFE** |

No caller is unresolved. Nothing rounds toward safe: the last row is called out as conditional rather than safe because its safety rests on a dtype invariant enforced elsewhere, not on any contiguity property at the call site.

## SSD spill verdict — not exposed, and deliberately so

This is the worst-case branch (corrupt bytes written to disk, hydrated later as valid, surviving the process), so it gets its own section. **It does not happen.** Two independent guards, both intentional.

**1. The prefill seed is explicitly materialised.** `exit_prefill` (`crates/rmlx-kv-quant/src/kvcache/update.rs:2686-2697`) stores the bf16 decode seed through `.contiguous()`, and the comment immediately above it (lines 2665-2681) names this exact hazard and this exact consumer:

> `contiguous` and not `try_clone`: `k_full` is a slice of the raw prefill buffer along the sequence axis, and MLX keeps that as a strided view over the parent allocation — `eval` does not flatten it. […] anything that reads the buffer by raw linear offset rather than by stride — `Array::to_bytes`, which is how the SSD tier serialises it — sees the parent's leading bytes under the slice's shape, i.e. head 0's whole row window in place of every head.

So the hazard is already known in-tree and was closed per-caller at the one call site that mattered. That is worth stating plainly: this report is not a new discovery of the mechanism, it is a census of everywhere else the mechanism reaches.

**2. The spill writer refuses a slab it cannot trust.** `none_bf16_payloads` (`crates/rmlx-kv-ssd/src/block_io.rs:352-378`) rejects the whole block when `c.offset() != k.shape()[2]`. After the first decode step the bf16 mirror is the full `[b, kv_h, max_seq, head_dim]` allocation (`update_decode_fp16_v_slab` stores the slab, not the prefix, at `update.rs:4581-4584`), so `shape[2] == max_seq != offset` and the block is refused rather than written. Costs a re-prefill; cannot corrupt.

The lead pointing at `update.rs:2888` as the persisted-spill feeder is **refuted**. That line is the `KvQuant::None` arm promoting `raw_k`/`raw_v` — the full prefill ring buffers, dense by construction from `zeros` + `slice_update`, not views. It is also shadowed by the `KvStorage::None` early return at `update.rs:2715`.

## V-slab path — not exposed today, but it is the loaded shape

`update_decode_fp16_v_only` (`update.rs:4589-4596`) returns `slice_v_prefix` of the `[b, kv_h, max_seq, head_dim]` slab — row-contiguous only when `b * kv_h == 1`, i.e. non-contiguous exactly when `kv_h > 1`. Its four callers (`update.rs:7146, 7197, 7673, 7741`) return it as `v_full` to the SDPA caller. It never reaches `to_bytes`; the flash-decode dispatchers take the mirror whole and stride over it via `flatten_v_mirror` (`crates/rmlx-kv-quant/src/flash_decode_common.rs:159`), which reshapes rather than reads bytes.

So: safe, but this is a non-contiguous rank-4 `Array` handed across a crate boundary as a plain value. Any future consumer that calls `to_bytes` on it gets silent corruption at `kv_h > 1` and correct results at `kv_h == 1` — which is to say, correct on gemma-4-e2b and wrong on Bonsai-8B, the exact split that single-architecture testing misses.

## Concrete failure mode — the conditionally-safe class

Take `update_iso_k_only_4` (`update.rs:7811-7816`) or any of the sibling K-only iso/rotor arms. Each decode step:

```rust
let v_f32 = array_to_f32_vec(new_v, Device::Cpu)?;
```

`new_v` arrives from the arch attention block as a transposed view. Every architecture in the tree does this (`qwen3.rs:1074`, `gemma3/attention.rs:108`, `qwen2/model.rs:293`, `qwen3_vl_moe/attention.rs:133`, `bitnet/model.rs:186`, `laguna/attention.rs`, `speculative/eagle3/mod.rs:490`):

```rust
let v = v.reshape(&[batch, seq, n_kv_heads, head_dim], device)?;
let v = v.transpose(&[0, 2, 1, 3], device)?;   // strided view
// K then goes through rope() — a real op, dense output.
// V does not. It reaches the cache as the view.
```

`KvCache::update` (`update.rs:1997-2043`) applies no cast and no `contiguous` before dispatching to the codec arms.

Three conditions must hold together for this to produce wrong values:

1. **`seq > 1`.** Measured above: at `seq == 1` the transpose is contiguous (the permuted axis has extent 1), at `seq > 1` it is not. Multi-token prefill routes to `update_prefill_raw`, which uses stride-aware `slice_update` and never reads bytes — safe. But **speculative decoding's verify step feeds `n_draft + 1` tokens outside the prefill bracket** (`speculative/mod.rs:1376-1405` brackets only the initial prompt), so `seq > 1` with `in_prefill == false` is reachable.
2. **A codec whose arm calls `array_to_f32_vec` on `new_v`.** Unconditional for the K-only iso/rotor and turbo families — the `Device::Cpu` argument there is the astype device, not a gate, so it runs every step on GPU too.
3. **`new_v.dtype() == Dtype::F32`.** The only remaining guard. Today the KV stream is bf16 (or f16 on PARO INT4 checkpoints), so `astype` fires and launders the view.

Condition 3 is the whole margin, and it is the one this tree has lost before — `make check-no-scalar-f32-leak` exists because f32 promotions into the KV stream recur. Note also that the bf16 floor (`cast_store_bf16`) is applied in `update_prefill_raw` only; the decode arms take `new_v` raw.

**What goes wrong if all three hold:** `to_bytes` returns the pre-transpose head-major bytes. Those are fed to `transpose_heads_seq(f32_data, b, kv_h, new_seq, head_dim)` (e.g. `storage/quant_iso_v.rs:257`), which reorders on the assumption that the vector is in `[b, kv_h, seq, d]` linear order. The length is right, so no check fires. The result is a quantized V store holding correctly-shaped, correctly-sized, wrong values. Downstream catches nothing: no length assert, no shape assert, no NaN. It surfaces as degraded generation quality on multi-token verify steps — attributable to the speculative drafter rather than to the codec, which is what makes it expensive to find. This is the same silent-corruption class as the out-of-bounds `slice_update` that froze a cache with no error.

## Gate status — two gates that cannot fail

No test anywhere reads `to_bytes` from a rank-≥2 non-contiguous view. The two that come closest both neutralise themselves:

- **`transpose_2d` — `crates/rmlx-mlx/src/lib_tests.rs:176`.** Builds a transposed view, then launders it through `matmul` and calls `to_bytes` on the matmul output. It asserts transpose's *semantics* via a dense result, and never reads the view's bytes. **Cannot fail on this hazard.**
- **`slice_1d` — `crates/rmlx-mlx/src/lib_tests.rs:212`.** Slices `[6] → [1:4]` and reads it back. Passes, and would pass under any implementation: a rank-1 stride-1 slice is contiguous by construction because the offset lands in the data pointer. It exercises the single slice shape the hazard does not affect. **Cannot fail on this hazard.**

`slice_update_1d` and `slice_update_2d_kv_pattern` read `slice_update` outputs, which are dense.

The fixture choice is what inverts the power here: `slice_1d` would become a real gate by adding one rank-2 case.

**Profile note.** There is no assertion in `to_bytes` today, so there is nothing that dies under `release-perf` — but any fix shaped as `debug_assert!` would be dead in exactly the profile that ships and that `make ci-perf`'s non-GPU half uses. A guard here must be a runtime check or a copy, never a `debug_assert`.

## Fix options

`mlx-c` 0.6.0_2 already exposes what a guard needs, and bindgen already emits all of it into `rmlx-mlx`'s generated bindings (verified present: `_mlx_array_is_contiguous`, `_mlx_array_is_row_contiguous`, `mlx_array_strides`). No new dependency, no new env var, no new config knob.

1. **`to_bytes` calls `.contiguous()` internally when the array is not row-contiguous.** Correct everywhere, no caller changes, costs a copy only on the views that would otherwise be wrong — and every such copy today is a copy that replaces a wrong answer. Needs a `Device` to relayout, which `to_bytes` does not currently take; that is the one wrinkle.
2. **`to_bytes` returns an error on non-row-contiguous input.** Loud, zero hot-path cost, forces each caller to decide. Turns a silent corruption into a startup-visible failure — but it is a behaviour change on a widely-called API, and it converts any missed caller from "works today" into "errors today".
3. **Guard only `array_to_f32_vec`'s F32 branch** (`helpers.rs:197-203`) — swap `try_clone()` for a contiguity check plus `contiguous()`. Smallest diff, covers the entire conditionally-safe class (all 66 helper sites), leaves the other 96 unguarded.
4. **Prove contiguity statically per caller.** This census is that exercise; it does not survive the next caller.

**Recommendation: (2) plus (3).** The census found no production caller that legitimately wants to read a strided view, so option 2's error is unreachable on today's code and pins the invariant loudly for every future caller at zero runtime cost — it makes the whole class a build-and-smoke-visible failure rather than a silent one. Option 3 then keeps the one call site that *can* legitimately receive a view working, by relaying out rather than erroring. Option 1 is the fallback if a caller that genuinely needs the copy turns up; it is strictly safer and its only cost is a copy that is currently a wrong answer.

Since the census is all-safe, an assertion would be defensible on its own — but the KV path already produces the hazardous shape on every decode step (the V transpose, the V slab), so the copy in option 3 is buying something real, not hypothetical.

## Required mutation check

A fix here is unfalsifiable without deliberately breaking it — the census being all-safe means no existing test changes colour. Whichever option lands:

1. Construct a strided view: `[1, 2, 4, 2]` f32, `.slice(&[0,0,0,0], &[1,2,2,2], &[1;4])`. Pass it through the fixed path. **Confirm the guard fires** — an error for option 2, a correct `[0,1,2,3,8,9,10,11]` for options 1/3.
2. Add the transpose case too: `[1,3,4,2].transpose(&[0,2,1,3])`, which reproduces the arch V shape at `seq > 1`.
3. **Mutation-check it**: revert the guard, confirm the new test goes red. A guard whose test passes before the fix is not a guard.
4. Extend `slice_1d` (`crates/rmlx-mlx/src/lib_tests.rs:212`) to a rank-2 slice, or add a sibling — as written it cannot fail.
5. **Run all of the above at `release-perf`, not only `dev`**: `make test-perf`, or `cargo test -p rmlx-mlx -p rmlx-kv-quant --profile release-perf`. `dev` alone would not catch a guard written as `debug_assert!`, and `release-perf` is what ships. These are CPU-only tests — no Metal context, no `#[ignore]`.

## Related

Issue #45 (fused flash-decode-over-quant) touches the same V-mirror surface. PR #469 reworked the V-slab area.
